toolchain/lower/handle.cpp

```diff
@@ -242,48 +251,70 @@ static auto HandleBuiltinCall(FunctionContext& context, SemIR::InstId inst_id,
     }
-    case SemIR::BuiltinFunctionKind::IntAdd: {
+    case SemIR::BuiltinFunctionKind::IntSAdd: {
       context.SetLocal(inst_id, context.builder().CreateAdd(
                                     context.GetValue(arg_ids[0]),
                                     context.GetValue(arg_ids[1]), "add",
                                     /*HasNUW=*/false,
                                     /*HasNSW=*/SignedOverflowIsUB));
       return;
     }
-    case SemIR::BuiltinFunctionKind::IntSub: {
+    case SemIR::BuiltinFunctionKind::IntSSub: {
       context.SetLocal(inst_id, context.builder().CreateSub(
                                     context.GetValue(arg_ids[0]),
                                     context.GetValue(arg_ids[1]), "sub",
                                     /*HasNUW=*/false,
                                     /*HasNSW=*/SignedOverflowIsUB));
       return;
     }
-    case SemIR::BuiltinFunctionKind::IntMul: {
+    case SemIR::BuiltinFunctionKind::IntSMul: {
       context.SetLocal(inst_id, context.builder().CreateMul(
                                     context.GetValue(arg_ids[0]),
                                     context.GetValue(arg_ids[1]), "mul",
                                     /*HasNUW=*/false,
                                     /*HasNSW=*/SignedOverflowIsUB));
       return;
     }
-    case SemIR::BuiltinFunctionKind::IntDiv: {
-      context.SetLocal(inst_id, IsSignedInt(context, inst_id)
-                                    ? context.builder().CreateSDiv(
-                                          context.GetValue(arg_ids[0]),
-                                          context.GetValue(arg_ids[1]), "div")
-                                    : context.builder().CreateUDiv(
-                                          context.GetValue(arg_ids[0]),
-                                          context.GetValue(arg_ids[1]), "div"));
+    case SemIR::BuiltinFunctionKind::IntSDiv: {
+      context.SetLocal(inst_id, context.builder().CreateSDiv(
+                                    context.GetValue(arg_ids[0]),
+                                    context.GetValue(arg_ids[1]), "div"));
       return;
     }
-    case SemIR::BuiltinFunctionKind::IntMod: {
-      context.SetLocal(inst_id, IsSignedInt(context, inst_id)
-                                    ? context.builder().CreateSRem(
-                                          context.GetValue(arg_ids[0]),
-                                          context.GetValue(arg_ids[1]), "rem")
-                                    : context.builder().CreateURem(
-                                          context.GetValue(arg_ids[0]),
-                                          context.GetValue(arg_ids[1]), "rem"));
+    case SemIR::BuiltinFunctionKind::IntSMod: {
+      context.SetLocal(inst_id, context.builder().CreateSRem(
+                                    context.GetValue(arg_ids[0]),
+                                    context.GetValue(arg_ids[1]), "rem"));
```

Comment:
Would it be helpful for names to reflect the builtin function rather than the call? e.g., "smod" (including signed versus unsigned) instead of "rem".

Reply:
Given that the builtin names are an API between the compiler and the core library, rather than user-facing, I don't know whether / how much a Carbon developer reading the LLVM IR will care about them matching; I think matching the name of the LLVM IR instruction might be more useful. But this all seems to connect to the broader question of naming; I think we should settle that before proceeding with this PR.

Reply:
FWIW, I thought these names would be used in the LLVM IR and mainly it might help match back operations to the thing that created them (at least, I thought that's what the purpose of these were). I was expecting the LLVM IR instruction to already be visible, as in:

```
// CHECK:STDOUT: define i32 @TestMod(i32 %a, i32 %b) {
// CHECK:STDOUT:   %rem = srem i32 %a, %b
// CHECK:STDOUT:   ret i32 %rem
// CHECK:STDOUT: }
```

->

```
// CHECK:STDOUT: define i32 @TestMod(i32 %a, i32 %b) {
// CHECK:STDOUT:   %smod = srem i32 %a, %b
// CHECK:STDOUT:   ret i32 %smod
// CHECK:STDOUT: }
```

Note, maybe this could even just pass the builtin name?

Reply:
I think the builtin name is an implementation detail of the core lib <-> compiler interface, and shouldn't be showing up in the IR -- except if, by "coincidence", it's also a good name for the result, which I'd expect it usually is.

But I do somewhat question how useful the instruction names we're adding here really are. It doesn't seem very valuable to me to add names that could be inferred from the IR itself, and giving instructions names does add some cost. Names like `%a` and `%b` above are clearly very useful, and give information that would not otherwise 
be present in the IR, but names that just describe the IR don't seem so useful to me.

Maybe we can just remove these names?

Reply:
Maybe they're not important -- I'll leave that to you. I'm not sure what the effect of just removing would be. You could also just call it "ret" in all cases. It seems like these are all short functions, so the value of naming the result is negligible.

FWIW %a %b could be %lhs %rhs in context.

Reply:
I've sent out #3898 to handle this more generally.